src/bottom-bar/complete-button.js

```diff
@@ -0,0 +1,75 @@
+import { useAlert } from '@dhis2/app-runtime'
+import i18n from '@dhis2/d2-i18n'
+import { Button } from '@dhis2/ui'
+import { useIsMutating } from '@tanstack/react-query'
+import PropTypes from 'prop-types'
+import React from 'react'
+import {
+    useConnectionStatus,
+    useDataValueSet,
+    useSetFormCompletionMutation,
+    useSetFormCompletionMutationKey,
+    useImperativeCancelCompletionMutation,
+} from '../shared/index.js'
+import useOnCompleteCallback from './use-on-complete-callback.js'
+
+const incompletingFormFailedMessage = i18n.t(
+    "Couldn't incomplete the form, please try again later"
+)
+
+export default function CompleteButton({ disabled }) {
+    const { offline } = useConnectionStatus()
+    const setFormCompletionMutationKey = useSetFormCompletionMutationKey()
+    const isLoading = useIsMutating(setFormCompletionMutationKey)
+    const { show: showErrorAlert } = useAlert((message) => message, {
+        critical: true,
+    })
+
+    const dataValueSet = useDataValueSet()
+    // `mutate` doesn't return a promise, `mutateAsync` does
+    const { mutateAsync: setFormCompletion } = useSetFormCompletionMutation()
+    const cancelCompletionMutation = useImperativeCancelCompletionMutation()
+
+    const onCompleteClick = useOnCompleteCallback()
+    const onIncompleteClick = function onIncompleteClick() {
+        if (isLoading) {
+            cancelCompletionMutation()
+        }
+
+        if (isLoading && offline) {
+            // There's no need to incomplete the form when it hasn't been
+            // completed yet. It's important to still perform the request
+            // online as we don't know if the mutation actually reached the server
+            // already
+            return
+        }
```

Comment:
I don't think there should be any difference between "incomplete" and "complete". This is just passed in as a variable to the mutation-function right?

You shouldn't handle these different, the previous state shouldn't really matter other than sending `complete: false` or `complete: true` to the mutate-fn.

Reply:
Yeah, the `useSetFormCompletionMutation` hook is re-used by both the `onCompleteClick` and the `onIncompleteClick` handlers.

The difference between them is which requests should be fired alongside. When incompleting, there's never another request being sent. When completing, then in two cases we also send the validation request, hence two different click handlers